common/src/main/proto/TransportMessages.proto

```diff
@@ -610,3 +612,24 @@ message PbTransportableError {
   int64 streamId = 1;
   string message = 2;
 }
+
+enum PbAuthType {
+  CLIENT_AUTH = 0;
+  CONNECTION_AUTH = 1;
+}
+
+message PbSaslMechanism {
+  string mechanism = 1;
+  repeated PbAuthType authTypes = 2;
+}
+
+message PbSaslRequest {
+  string appId = 1;
```

Comment:
This was part of `RegisterApplicationRequest`, why are we pulling it into SASL Request itself ?

Reply:
It has to be in both. The server needs to look up the secret and appId is the key.

Reply:
I am missing on why it needs to be in both.

During SASL, server is trying to authenticate the incoming client - which does not necessarily require an appId (for example, if we extend it to s2s connections, there is no appId).

If appId is being sent for identifying the client, that will be part of the `username` in SASL, and does not need to be part of the wrapper SASL message itself.

Spark handles it differently since it is specific to DIGEST-MD5 and assumes an app-id based auth.

Reply:
@mridulm This is how it is used in `SaslRpcHandler` 
https://github.com/apache/incubator-celeborn/pull/2105/files#diff-89b6059b671d6541a463356e31d4b6950ee21890ba3567a05aab19e58eeffb54

- Identify whether that the app has registered or not
- The appId in the first Sasl message of the handshake sets up the Sasl server which the server uses to retrieve the secret.

Reply:
The following example illustrates how to get this done without providing the login id out of band (typical sasl usages do not provide this for example):

(Culled from a scala worksheet, so not pretty :) )
```

import javax.security.auth.callback._
import javax.security.sasl._

val EMPTY_BYTE_ARRAY = Array[Byte]()
val USER = "my-user"

// proxy for password store ...
val passwordStore = Map(
  USER -> "my-password"
)

class MyServerCallback() extends CallbackHandler {

  private var userName: String = _

  override def handle(callbacks: Array[Callback]): Unit = {
    callbacks.foreach {
      case nc: NameCallback =>
        userName = Option(nc.getName).getOrElse(nc.getDefaultName)
      case pc: PasswordCallback =>
        val password = passwordStore.getOrElse(userName,
          throw new IllegalArgumentException(s"User $userName not found"))
        pc.setPassword(password.toCharArray)
      case rc: RealmCallback =>
        // ignore for now
      case rc: RealmChoiceCallback =>
      // ignore for now
      case ac: AuthorizeCallback =>
        // simple impl: for now, both match ...
        val authId = ac.getAuthenticationID
        val authzId = ac.getAuthorizationID
        ac.setAuthorized(authId == authzId)
        if (ac.isAuthorized) ac.setAuthorizedID(authzId)
      case c => throw new UnsupportedOperationException(s"Unsupported callback $c")
    }
  }
}

class MyClientCallback extends CallbackHandler {
  override def handle(callbacks: Array[Callback]): Unit = {
    callbacks.foreach {
      case nc: NameCallback =>
        nc.setName(USER)
      case pc: PasswordCallback =>
        pc.setPassword(passwordStore(USER).toCharArray)
      case rc: RealmCallback =>
        rc.setText(rc.getDefaultText)
      case rc: RealmChoiceCallback =>
        // ignore
      case c => throw new UnsupportedOperationException(s"Unsupported callback $c")
    }
  }
}

val saslProps = new java.util.HashMap[String, String]()
saslProps.put(Sasl.SERVER_AUTH, "true")


val saslServer = Sasl.createSaslServer("DIGEST-MD5", null, "default", saslProps,
  new MyServerCallback())

val saslClient = Sasl.createSaslClient(Array("DIGEST-MD5"), USER, null, "default", saslProps,
  new MyClientCallback())

var dataFromClient = {
  if (saslClient.hasInitialResponse) {
    saslClient.evaluateChallenge(EMPTY_BYTE_ARRAY)
  } else {
    EMPTY_BYTE_ARRAY
  }
}

while (!saslClient.isComplete) {
  val serverResult = saslServer.evaluateResponse(dataFromClient)
  dataFromClient = saslClient.evaluateChallenge(serverResult)
}

println("Auth success ! Client qop = " + saslClient.getNegotiatedProperty(Sasl.QOP))

saslClient.dispose()
saslServer.dispose()
```

Reply:
With this we are making an implicit assumptions on the server about the sasl-user - that the client will set it to application Id cause it has to be unique. In Spark's implementation, the `applicationId -> secret` mapping is explicit. Spark does not make any implicit assumption that the sasl-user is applicationId and I think that is better. 

With s2s there is no appId but then I can rename this field to `id`. For client to server, it will be set as applicationId. For s2s, it will be unique worker id.

Reply:
Spoke to @mridulm offline. Will change this to what Mridul suggested as it is the standard way to implement Sasl handshake.

Reply:
@mridulm I have updated the code. Please take a look.